src/main/java/io/cdap/plugin/gcp/dataplex/source/DataplexBatchSource.java

```diff
@@ -0,0 +1,514 @@
+      emitter.emit(transformed);
+    } else {
+      emitter.emit((StructuredRecord) input.getValue());
+    }
+  }
+
+  /**
+   * Gets bucket from supplied configuration.
+   * <p>
+   * If the supplied configuration doesn't specify a bucket, a bucket will get auto created and configuration modified
+   * to auto-delete this bucket on completion.
+   *
+   * @param configuration Hadoop configuration instance.
+   * @param project       GCP projectId
+   * @param bigQuery      bigquery client
+   * @param credentials   GCP credentials
+   * @param bucketName    bucket Name
+   * @return Bucket name.
+   */
+  private String createBucket(Configuration configuration,
+                              String project,
+                              BigQuery bigQuery,
+                              Credentials credentials,
+                              String bucketName) {
+    String bucket = String.format(BQ_TEMP_BUCKET_NAME_TEMPLATE, bucketName);
+    // By default, this option is false, meaning the job can not delete the bucket.
+    configuration.setBoolean("fs.gs.bucket.delete.enable", true);
+
+    // the dataset existence is validated before, so this cannot be null
+    Dataset bigQueryDataset = bigQuery.getDataset(DatasetId.of(datasetProject, dataset));
+    GCPUtils.createBucket(GCPUtils.getStorage(project, credentials),
+      bucket, bigQueryDataset.getLocation(), null);
+    return bucket;
+  }
+
+  /**
+   * Sets up service account credentials into supplied Hadoop configuration.
+   *
+   * @param configuration Hadoop Configuration instance.
+   * @param config        BigQuery connection configuration.
+   */
+  private void configureServiceAccount(Configuration configuration, GCPConnectorConfig config) {
+    if (config.getServiceAccount() != null) {
+      configuration.set(BigQueryConstants.CONFIG_SERVICE_ACCOUNT, config.getServiceAccount());
+      configuration.setBoolean(BigQueryConstants.CONFIG_SERVICE_ACCOUNT_IS_FILE, config.isServiceAccountFilePath());
+    }
+  }
+
+  /**
+   * Gets bucket from supplied configuration.
+   * <p>
+   * If the supplied configuration doesn't specify a bucket, a bucket will get auto created and configuration modified
+   * to auto-delete this bucket on completion.
+   *
+   * @param configuration Hadoop configuration instance.
+   * @param storage       GCS Storage client
+   * @param bucket        GCS bucket
+   * @param location      location
+   * @return Bucket name.
+   */
+  private String createBucket(Configuration configuration,
+                              Storage storage, String location,
+                              @Nullable String bucket) throws IOException {
+    // Create a new bucket if needed
+    if (storage != null && storage.get(bucket) == null) {
+      try {
+        configuration.setBoolean("fs.gs.bucket.delete.enable", true);
+        GCPUtils.createBucket(storage, bucket, location, null);
```

Comment:
similar comment about CMEK here as well.

Reply:
will be added later